File: mediaProjectionLibrary/src/main/java/tech/thdev/media_projection_library/view/MediaProjectionAccessActivity.java
package tech.thdev.media_projection_library.view;

import android.content.Context;
import android.content.Intent;
import android.media.projection.MediaProjectionManager;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import tech.thdev.media_projection_library.param.MediaProjectionControllerParams;
import tech.thdev.media_projection_library.presenter.MediaProjectionAccessPresenter;
import tech.thdev.media_projection_library.presenter.view.MediaProjectionAccessView;

/**
 * Created by Tae-hwan on 4/25/16.
 * <p/>
 * MediaProjection Abstract
 */
public abstract class MediaProjectionAccessActivity extends AppCompatActivity implements MediaProjectionAccessView {

    public static final int REQ_CODE_MEDIA_PROJECTION = 1000;

    private MediaProjectionAccessPresenter mediaProjectionAccessPresenter;
    private MediaProjectionManager mediaProjectionManager;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mediaProjectionManager = (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
        mediaProjectionAccessPresenter = new MediaProjectionAccessPresenter(this, this, mediaProjectionManager);
    }

    /**
     * Projection start
     *
     * @param controllerParams
     */
    public void createMediaProjection(MediaProjectionControllerParams controllerParams) {
        mediaProjectionAccessPresenter.setControllerParams(controllerParams);
        startActivityForResult(mediaProjectionManager.createScreenCaptureIntent(), REQ_CODE_MEDIA_PROJECTION);
    }

    /**
     * MediaProjection stop.
     */
    public void stopMediaProjection() {
        mediaProjectionAccessPresenter.destroyMediaProjection();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case REQ_CODE_MEDIA_PROJECTION:
                mediaProjectionAccessPresenter.initMediaProjection(resultCode, data);
                break;

            default:
                super.onActivityResult(requestCode, resultCode, data);
                break;
        }
    }
}

File: mediaProjectionLibrary/src/main/java/tech/thdev/media_projection_library/ui/MediaProjectionAccessActivity.kt
package tech.thdev.media_projection_library.ui

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.media.projection.MediaProjectionManager
import android.os.Bundle
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContract
import androidx.appcompat.app.AppCompatActivity
import tech.thdev.media_projection_library.R

/**
 * Created by Tae-hwan on 4/25/16.
 * @since 4/12/20 Edit kotlin.
 *
 * MediaProjection mediaProjection manager
 */
internal class MediaProjectionAccessActivity : AppCompatActivity() {

    companion object {
        private const val REQ_CODE_MEDIA_PROJECTION = 1000

        fun newInstance(context: Context): Intent =
            Intent(context, MediaProjectionAccessActivity::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
            }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_media_projection)
        createMediaProjection()
    }

    private fun createMediaProjection() {
        val mediaProjectionManager = getSystemService(Context.MEDIA_PROJECTION_SERVICE) as MediaProjectionManager
        val result: ActivityResultLauncher<Intent> = registerForActivityResult(MediaProjectionResultContract()) {
            sendBroadcast(it)
            finish()
        }
        result.launch(mediaProjectionManager.createScreenCaptureIntent())

        // Old forResult
//        startActivityForResult(
//            mediaProjectionManager.createScreenCaptureIntent(),
//            REQ_CODE_MEDIA_PROJECTION
//        )
    }

//    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
//        when (requestCode) {
//            REQ_CODE_MEDIA_PROJECTION -> sendBroadcast(resultCode, data)
//            else -> super.onActivityResult(requestCode, resultCode, data)
//        }
//    }
}

internal class MediaProjectionResultContract : ActivityResultContract<Intent, Intent>() {

    override fun createIntent(context: Context, input: Intent?): Intent =
        input!!

    override fun parseResult(resultCode: Int, intent: Intent?): Intent {
        return if (resultCode == Activity.RESULT_OK && intent != null) {
            MediaProjectionBroadcastReceiver.newInstance(resultCode, intent)
        } else {
            MediaProjectionBroadcastReceiver.newReject()
        }
    }
}